tunnel_test.go

```diff
@@ -0,0 +1,62 @@
+package k8sportforward
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/fortytw2/leaktest"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
+)
+
+type fakeDialer struct {
+	dialed             bool
+	conn               httpstream.Connection
+	err                error
+	negotiatedProtocol string
+}
+
+func (d *fakeDialer) Dial(protocols ...string) (httpstream.Connection, string, error) {
+	d.dialed = true
+	return d.conn, d.negotiatedProtocol, d.err
+}
+
+func TestGoroutineLeak(t *testing.T) {
+	defer leaktest.Check(t)()
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintln(w, "Hello, client")
+	}))
+	defer ts.Close()
+	parsedURL, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatalf("could not parse test server URL %v", err)
+	}
+
+	conn, err := net.Dial("tcp", parsedURL.Host)
+	if err != nil {
+		t.Fatalf("could not create connection to server %v", err)
+	}
+
+	tunnelCfg := TunnelConfig{
+		Remote:    80,
+		Namespace: "default",
+		PodName:   "test",
+	}
+
+	streamConn, err := spdy.NewClientConnection(conn)
+	if err != nil {
+		t.Fatalf("could not create connection to server %v", err)
+	}
+	dialer := &fakeDialer{
+		conn: streamConn,
+	}
+	_, err = newTunnel(dialer, tunnelCfg)
```

Comment:
How is the tunnel closed after it is established and we verified what we want to test? How does the test finally end?

Reply:
With this setup the tunnel is closed after `ts` server is closed. 

Anyway we are not having issues with not closed tunnels, the point of the test is checking that this goroutine https://github.com/giantswarm/k8sportforward/blob/master/tunnel.go#L89-L91 is not leaked anymore, after the analysis with pprof described here https://github.com/giantswarm/giantswarm/issues/3607 it turned out that it is the source of the problem. 

As stated above, going back to the old goroutine definition makes the test fail.

Reply:
What makes the test server closing then?

Reply:
> What makes the test server closing then?

`defer ts.Close()` here https://github.com/giantswarm/k8sportforward/pull/7/files#diff-b247ac39dcfc87620ab8cb858fc0e663R34

Reply:
Oh I see. With you.